### PR TITLE
Add Compatibility matrix version check

### PR DESCRIPTION
**Pull request by asankov-cb**

### CODE REVIEW

File: cbcontainers/communication/gateway/api_gateway.go [hunk 1]
```diff
@@ -4,10 +4,16 @@ import (
...
 	"fmt"
+	"net/http"
+
 	"github.com/go-resty/resty/v2"
 	"github.com/vmware/cbcontainers-operator/cbcontainers/models"
-	"net/http"
+)
+
+var (
+	ErrGettingCompatibilityMatrix = errors.New("error while getting compatibility matrix")
```
Comment (RazOmessi) [inline cbcontainers/communication/gateway/api_gateway.go:16]: I think the error can be change a bit cause we are not getting the matrix anymore

File: cbcontainers/models/compatibility_matrix.go [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+func (v AgentVersion) IsLessThan(version string) bool {
+	if v == AgentMaxVersionLatest || v == AgentVersionUnknown {
+		return false
+	}
+	return string(v) < version
+}
+
+// CompatibilityMatrixResponse is the response returned by the GET Compatibility Matrix API.
+type CompatibilityMatrixResponse struct {
```
Comment (RazOmessi) [inline cbcontainers/models/compatibility_matrix.go:28]: can be deleted?

File: cbcontainers/models/compatibility_matrix.go [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+	Operators CompatibilityMatrix `json:"operators"`
+}
+
+type CompatibilityMatrix map[string]*CompatibilityMatrixEntry
```
Comment (RazOmessi) [inline cbcontainers/models/compatibility_matrix.go:32]: same

File: cbcontainers/models/compatibility_matrix.go [hunk 4]
```diff
@@ -0,0 +1,55 @@
+package models
```
Comment (RazOmessi) [inline cbcontainers/models/compatibility_matrix.go:1]: I think I would change the file name to `CompatibilityEntry` or something like this.
And maybe add another file to the `AgentVersion`

File: controllers/cbcontainerscluster_controller.go [hunk 5]
```diff
@@ -43,11 +43,32 @@ type ClusterProcessor interface {
...
+	GetOperatorVersion() (string, error)
+}
+
 type CBContainersClusterReconciler struct {
 	client.Client
 	Log    logr.Logger
 	Scheme *runtime.Scheme
 
+	// GatewayCreator will be used to create a gateway.ApiGateway
+	GatewayCreator
```
Comment (RazOmessi) [inline controllers/cbcontainerscluster_controller.go:64]: I think i would not put this as embed field

File: controllers/cbcontainerscluster_controller.go [hunk 6]
```diff
@@ -43,11 +43,32 @@ type ClusterProcessor interface {
...
+	// _apiGateway is a singleton gateway.ApiGateway struct.
+	// It is initialized the first time it is used and reused after that.
+	// It should not be accessed directly, but through the apiGateway method.
+	_apiGateway Gateway
+
+	OperatorVersionProvider
```
Comment (RazOmessi) [inline controllers/cbcontainerscluster_controller.go:70]: same

File: controllers/cbcontainerscluster_controller.go [hunk 7]
```diff
@@ -115,6 +159,13 @@ func (r *CBContainersClusterReconciler) Reconcile(ctx context.Context, req ctrl.
 	return ctrl.Result{Requeue: stateWasChanged}, nil
 }
 
+func (r *CBContainersClusterReconciler) apiGateway(cbContainersCluster *cbcontainersv1.CBContainersCluster, accessToken string) Gateway {
+	if r._apiGateway == nil {
```
Comment (RazOmessi) [inline controllers/cbcontainerscluster_controller.go:163]: Im not sure if it can happen but:

if some one will change something in the cbContainersCluster or the accessToken and we already created the _apiGateway, we might return here invalid gateway

**Review by RazOmessi — CHANGES_REQUESTED** [review]

File: cbcontainers/models/compatibility_matrix.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+type AgentVersion string
+
+const (
+	AgentMinVersionNone   AgentVersion = "none"
+	AgentMaxVersionLatest AgentVersion = "latest"
+	AgentVersionUnknown   AgentVersion = ""
+)
+
+func (v AgentVersion) IsBiggerThan(version string) bool {
```
Comment (ltsonov-cb) [inline cbcontainers/models/compatibility_matrix.go:13]: Shouldn't `latest` be bigger than all ? 

Also I think `Larger` is more appropriate term (less <-> large, small <-> big)

File: cbcontainers/models/compatibility_matrix.go [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+	if v == AgentMinVersionNone || v == AgentVersionUnknown {
+		return false
+	}
+	return string(v) > version
+}
+
+func (v AgentVersion) IsLessThan(version string) bool {
+	if v == AgentMaxVersionLatest || v == AgentVersionUnknown {
```
Comment (ltsonov-cb) [inline cbcontainers/models/compatibility_matrix.go:21]: As above, should `none` be smaller than all?

File: cbcontainers/models/compatibility_matrix.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+// It returns an error if the versions are not compatible, nil if they are.
+// If the min and max agent versions for this operator version are unknown it will
+// skip the check and return true.
+func (entry CompatibilityMatrixEntry) CheckCompatibility(agentVersion string) error {
+	if entry.Max.IsLessThan(agentVersion) {
+		return fmt.Errorf("agent version too high, upgrade the operator to use that agent version")
+	}
+	if entry.Min.IsBiggerThan(agentVersion) {
+		return fmt.Errorf("agent version too low, downgrade the operator to use that agent version")
+	}
```
Comment (ltsonov-cb) [inline cbcontainers/models/compatibility_matrix.go:51]: We could include the actual min/max version required in the error to make the consumer's life easier (i.e. they don't have to reproduce the steps to check the compatibility - they can directly see it in the error)

File: cbcontainers/models/compatibility_matrix_test.go [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/vmware/cbcontainers-operator/cbcontainers/models"
+)
+
+func TestCheckCompatibility(t *testing.T) {
```
Comment (ltsonov-cb) [inline cbcontainers/models/compatibility_matrix_test.go:11]: I think this should be 2 tests - one that verify compatible agents are detected and one that detects incompatible agents are detected

File: cbcontainers/models/compatibility_matrix_test.go [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+		agent         string
+		areCompatible bool
+	}{
+		{min: models.AgentVersion("2.7"), max: models.AgentVersion("2.8"), agent: "2.7", areCompatible: true},
+		{min: models.AgentVersion("2.7"), max: models.AgentVersion("2.8"), agent: "2.8", areCompatible: true},
+		{min: models.AgentVersion("2.7"), max: models.AgentVersion("2.8"), agent: "2.7.1", areCompatible: true},
+		{min: models.AgentVersion("2.7"), max: models.AgentVersion("2.8"), agent: "2.7.5", areCompatible: true},
+		{min: models.AgentVersion("2.7"), max: models.AgentVersion("2.8.1"), agent: "2.8.1", areCompatible: true},
+		{min: models.AgentVersion("2.7"), max: models.AgentVersion("2.8.1"), agent: "2.7.1-alpha", areCompatible: true},
+		{min: models.AgentVersion("2.7"), max: models.AgentVersion("2.8.1"), agent: "2.7.1-beta", areCompatible: true},
```
Comment (ltsonov-cb) [inline cbcontainers/models/compatibility_matrix_test.go:24]: Nice catch to include those test cases

File: cbcontainers/models/compatibility_matrix_test.go [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+		{min: models.AgentVersion("0.0"), max: models.AgentVersion("2.8"), agent: "0.0", areCompatible: true},
+		{min: models.AgentVersion("0.0"), max: models.AgentVersion("2.8"), agent: "1.0", areCompatible: true},
+		{min: models.AgentVersion("0.0"), max: models.AgentVersion("2.8"), agent: "2.0", areCompatible: true},
+
+		{min: models.AgentMinVersionNone, max: models.AgentMaxVersionLatest, agent: "0.0", areCompatible: true},
+		{min: models.AgentMinVersionNone, max: models.AgentMaxVersionLatest, agent: "0.0.1", areCompatible: true},
+		{min: models.AgentMinVersionNone, max: models.AgentMaxVersionLatest, agent: "0.0.0-beta", areCompatible: true},
+		{min: models.AgentMinVersionNone, max: models.AgentMaxVersionLatest, agent: "30.0.0-beta", areCompatible: true},
+
+		{min: models.AgentVersion("2.7"), max: models.AgentVersion("2.8"), agent: "2.8.1", areCompatible: false},
```
Comment (ltsonov-cb) [inline cbcontainers/models/compatibility_matrix_test.go:39]: An interesting test case here is 
`minVersion=2.7.0, agentVersion=2.7.0-beta`

Which requires that we decide whether a beta/alpha is < than the actual version. In the test cases above we treat X.Y-beta > X.(Y-1) but it is not clear whether X.Y-beta < X.Y 

I think this is a very edge cases and I am OK with ignoring it

File: controllers/cbcontainerscluster_controller.go [hunk 14]
```diff
@@ -115,6 +159,13 @@ func (r *CBContainersClusterReconciler) Reconcile(ctx context.Context, req ctrl.
...
+		r._apiGateway = r.GatewayCreator.CreateGateway(cbContainersCluster, accessToken)
+	}
+	return r._apiGateway
```
Comment (ltsonov-cb) [inline controllers/cbcontainerscluster_controller.go:166]: I think you need a lock here if you want to treat `_apiGateway` as a singleton.
2 routines can run in this way, using 2 instances in practice

```
R1 runs line 163, enters if
R2 runs line 163, enters if
R1 runs line 164, assigns _apigateway
R1 runs line 166, returns _apigateway (1)
R2 runs line 164, assigns _apigateway again
R2 runs line 166, returns _apigateway(2)
```


On a separate note - what are the pros of this being a singleton - is it expensive to create ? 

As 1 con, this will essentially lock the instance to use the same access token indefinitely, which might not be good long-term if tokens need refreshing (e.g. if we move to OAuth or similar).

One idea could be to use 1 instance of the HTTP client but make the authentication per request - so that you create the expensive parts once but set the dynamic parts (e.g. Auth) every time. PSCKit's http client uses a similar approach and I think it is elegant.

File: controllers/cbcontainerscluster_controller_test.go [hunk 15]
```diff
@@ -173,12 +194,56 @@ func TestClusterReconcile(t *testing.T) {
...
 		require.Equal(t, result, ctrlRuntime.Result{})
 	})
+
+	t.Run("When it gets the compatibility matrix succesfully it should make a compatibility check", func(t *testing.T) {
+		t.Run("When agent version is between min and max it should not return an error", func(t *testing.T) {
+			result, err := testCBContainersClusterController(t, setupClusterCustomResource, setUpTokenSecretValues, func(testMocks *ClusterControllerTestMocks) {
+				testMocks.gatewayCreator.EXPECT().CreateGateway(gomock.Any(), gomock.Any()).Return(testMocks.gateway)
+
+				operatorVersion := "3.1.0"
+				testMocks.operatorVersionProvider.EXPECT().GetOperatorVersion().Return(operatorVersion, nil)
```
Comment (ltsonov-cb) [inline controllers/cbcontainerscluster_controller_test.go:212]: [Minor] You could use the value provided by `setupOperatorVersionProvider` and utilize `gomock.Any()` when setting up the compatibility matrix mock? Since the operator version is essentially meaningless to this test, the agent version is what matters.

File: cbcontainers/models/compatibility_matrix.go (hunk 8 above)
Comment (asankov-cb) [reply cbcontainers/models/compatibility_matrix.go:13]: in this context we don't care about `latest`. we only care if `v` (which is the minimum) is bigger/larger than `version` which means we trying to install too old agent version

File: controllers/cbcontainerscluster_controller_test.go (hunk 15 above)
Comment (asankov-cb) [reply controllers/cbcontainerscluster_controller_test.go:212]: I believe that is wrong. `GetCompatibilityMatrixEntryFor` should be called with whatever comes from `operatorVersionProvider.GetOperatorVersion`. the test currently verifies that. if I use `gomock.Any()` it will not verify that and it will be less reliable

File: controllers/cbcontainerscluster_controller.go (hunk 14 above)
Comment (asankov-cb) [reply controllers/cbcontainerscluster_controller.go]: this singleton bring more harm than good. I removed it

File: cbcontainers/models/compatibility_matrix_test.go (hunk 11 above)
Comment (asankov-cb) [reply cbcontainers/models/compatibility_matrix_test.go]: the tests shared some code, that's why I made them this way. I extracted them into two top level tests, but they share a method that runs the logic

File: cbcontainers/models/compatibility_matrix.go (hunk 10 above)
Comment (asankov-cb) [reply cbcontainers/models/compatibility_matrix.go]: yes, that is a good idea. done.

File: controllers/cbcontainerscluster_controller.go (hunk 7 above)
Comment (asankov-cb) [reply controllers/cbcontainerscluster_controller.go]: it absolutely can. I removed the singleton. creating the gateway should not be an expensive operation. it just sets some struct fields.

File: Makefile [hunk 16]
```diff
@@ -120,7 +120,7 @@ generate: controller-gen
 
 # Build the docker image
 docker-build:
-	docker build -t ${IMG} .
+	docker build -t ${IMG} --build-arg OPERATOR_VERSION=${VERSION} .
```
Comment (ltsonov-cb) [inline Makefile:123]: This will only be applied when building the docker image but not when debugging locally  via `make run` - maybe we can fix that by setting the variable when doing `go run` as well?

File: cbcontainers/communication/gateway/api_gateway.go [hunk 17]
```diff
@@ -101,3 +109,23 @@ func (gateway *ApiGateway) GetCertificates(name string, privateKey *rsa.PrivateK
 
 	return gateway.getCertificates(commonName, organizationalUnit, organization, privateKey)
 }
+
+func (gateway *ApiGateway) GetCompatibilityMatrixEntryFor(operatorVersion string) (*models.OperatorCompatibility, error) {
+	url := gateway.baseUrl("setup/compatibility")
+	resp, err := gateway.baseRequest().
+		SetResult(&models.OperatorCompatibility{}).
+		SetQueryParam("operator_version", operatorVersion).
```
Comment (ltsonov-cb) [inline cbcontainers/communication/gateway/api_gateway.go:117]: Should be a path param

**Comment by asankov-cb** [comment]
closing in favour of https://github.com/octarinesec/octarine-operator/pull/60 which is rebased onto master and all conflicts are resolved
